## Import Libraries 

In [1]:
import pandas as pd
import streamlit as st
import plotly.express as px
from src.data import eda_func as fc
from src.data import px_func as pc

## Reading Data Set

In [2]:
animes_df = pd.read_csv('anime.csv')

In [3]:
pd.set_option('display.max_columns', None)

## EDA

In [4]:
animes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17562 entries, 0 to 17561
Data columns (total 35 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   MAL_ID         17562 non-null  int64 
 1   Name           17562 non-null  object
 2   Score          17562 non-null  object
 3   Genres         17562 non-null  object
 4   English name   17562 non-null  object
 5   Japanese name  17562 non-null  object
 6   Type           17562 non-null  object
 7   Episodes       17562 non-null  object
 8   Aired          17562 non-null  object
 9   Premiered      17562 non-null  object
 10  Producers      17562 non-null  object
 11  Licensors      17562 non-null  object
 12  Studios        17562 non-null  object
 13  Source         17562 non-null  object
 14  Duration       17562 non-null  object
 15  Rating         17562 non-null  object
 16  Ranked         17562 non-null  object
 17  Popularity     17562 non-null  int64 
 18  Members        17562 non-n

In [5]:
animes_df.head()

MAL_ID                             Name Score  \
0       1                     Cowboy Bebop  8.78   
1       5  Cowboy Bebop: Tengoku no Tobira  8.39   
2       6                           Trigun  8.24   
3       7               Witch Hunter Robin  7.27   
4       8                   Bouken Ou Beet  6.98   

                                              Genres            English name  \
0    Action, Adventure, Comedy, Drama, Sci-Fi, Space            Cowboy Bebop   
1              Action, Drama, Mystery, Sci-Fi, Space  Cowboy Bebop:The Movie   
2  Action, Sci-Fi, Adventure, Comedy, Drama, Shounen                  Trigun   
3  Action, Mystery, Police, Supernatural, Drama, ...      Witch Hunter Robin   
4          Adventure, Fantasy, Shounen, Supernatural  Beet the Vandel Buster   

                      Japanese name   Type Episodes  \
0                         カウボーイビバップ     TV       26   
1                    カウボーイビバップ 天国の扉  Movie        1   
2                             トライガン     TV       26   
3  Witch Hunter ROBIN (ウイッチハンターロビン)     TV       26   
4                            冒険王ビィト     TV       52   

                          Aired    Premiered  \
0   Apr 3, 1998 to Apr 24, 1999  Spring 1998   
1                   Sep 1, 2001      Unknown   
2   Apr 1, 1998 to Sep 30, 1998  Spring 1998   
3   Jul 2, 2002 to Dec 24, 2002  Summer 2002   
4  Sep 30, 2004 to Sep 29, 2005    Fall 2004   

                                           Producers  \
0                                      Bandai Visual   
1                             Sunrise, Bandai Visual   
2                               Victor Entertainment   
3  TV Tokyo, Bandai Visual, Dentsu, Victor Entert...   
4                                   TV Tokyo, Dentsu   

                              Licensors         Studios    Source  \
0      Funimation, Bandai Entertainment         Sunrise  Original   
1           Sony Pictures Entertainment           Bones  Original   
2  Funimation, Geneon Entertainment USA        Madhouse     Manga   
3      Funimation, Bandai Entertainment         Sunrise  Original   
4                               Unknown  Toei Animation     Manga   

          Duration                          Rating  Ranked  Popularity  \
0  24 min. per ep.  R - 17+ (violence & profanity)    28.0          39   
1    1 hr. 55 min.  R - 17+ (violence & profanity)   159.0         518   
2  24 min. per ep.       PG-13 - Teens 13 or older   266.0         201   
3  25 min. per ep.       PG-13 - Teens 13 or older  2481.0        1467   
4  23 min. per ep.                   PG - Children  3710.0        4369   

   Members  Favorites  Watching  Completed  On-Hold  Dropped  Plan to Watch  \
0  1251960      61971    105808     718161    71513    26678         329800   
1   273145       1174      4143     208333     1935      770          57964   
2   558913      12944     29113     343492    25465    13925         146918   
3    94683        587      4300      46165     5121     5378          33719   
4    13224         18       642       7314      766     1108           3394   

   Score-10   Score-9   Score-8  Score-7  Score-6 Score-5 Score-4 Score-3  \
0  229170.0  182126.0  131625.0  62330.0  20688.0  8904.0  3184.0  1357.0   
1   30043.0   49201.0   49505.0  22632.0   5805.0  1877.0   577.0   221.0   
2   50229.0   75651.0   86142.0  49432.0  15376.0  5838.0  1965.0   664.0   
3    2182.0    4806.0   10128.0  11618.0   5709.0  2920.0  1083.0   353.0   
4     312.0     529.0    1242.0   1713.0   1068.0   634.0   265.0    83.0   

  Score-2 Score-1  
0   741.0  1580.0  
1   109.0   379.0  
2   316.0   533.0  
3   164.0   131.0  
4    50.0    27.0

In [6]:
animes_df.duplicated().sum()

0

In [7]:
# Converting the Score, Score-10, Score-1 and Episodes columns to int/float
fc.convert_scores(animes_df)

MAL_ID                             Name  Score  \
0           1                     Cowboy Bebop   8.78   
1           5  Cowboy Bebop: Tengoku no Tobira   8.39   
2           6                           Trigun   8.24   
3           7               Witch Hunter Robin   7.27   
4           8                   Bouken Ou Beet   6.98   
...       ...                              ...    ...   
17557   48481  Daomu Biji Zhi Qinling Shen Shu    NaN   
17558   48483                     Mieruko-chan    NaN   
17559   48488  Higurashi no Naku Koro ni Sotsu    NaN   
17560   48491      Yama no Susume: Next Summit    NaN   
17561   48492                    Scarlet Nexus    NaN   

                                                  Genres  \
0        Action, Adventure, Comedy, Drama, Sci-Fi, Space   
1                  Action, Drama, Mystery, Sci-Fi, Space   
2      Action, Sci-Fi, Adventure, Comedy, Drama, Shounen   
3      Action, Mystery, Police, Supernatural, Drama, ...   
4              Adventure, Fantasy, Shounen, Supernatural   
...                                                  ...   
17557                   Adventure, Mystery, Supernatural   
17558                       Comedy, Horror, Supernatural   
17559  Mystery, Dementia, Horror, Psychological, Supe...   
17560                   Adventure, Slice of Life, Comedy   
17561                                    Action, Fantasy   

                          English name                     Japanese name  \
0                         Cowboy Bebop                         カウボーイビバップ   
1               Cowboy Bebop:The Movie                    カウボーイビバップ 天国の扉   
2                               Trigun                             トライガン   
3                   Witch Hunter Robin  Witch Hunter ROBIN (ウイッチハンターロビン)   
4               Beet the Vandel Buster                            冒険王ビィト   
...                                ...                               ...   
17557                          Unknown                         盗墓笔记之秦岭神树   
17558                          Unknown                           見える子ちゃん   
17559  Higurashi:When They Cry – SOTSU                        ひぐらしのなく頃に卒   
17560                          Unknown                ヤマノススメ Next Summit   
17561                          Unknown                     SCARLET NEXUS   

        Type  Episodes                         Aired    Premiered  \
0         TV        26   Apr 3, 1998 to Apr 24, 1999  Spring 1998   
1      Movie         1                   Sep 1, 2001      Unknown   
2         TV        26   Apr 1, 1998 to Sep 30, 1998  Spring 1998   
3         TV        26   Jul 2, 2002 to Dec 24, 2002  Summer 2002   
4         TV        52  Sep 30, 2004 to Sep 29, 2005    Fall 2004   
...      ...       ...                           ...          ...   
17557    ONA         0              Apr 4, 2021 to ?      Unknown   
17558     TV         0                     2021 to ?      Unknown   
17559     TV         0                Jul, 2021 to ?  Summer 2021   
17560     TV         0                       Unknown      Unknown   
17561     TV         0                Jul, 2021 to ?  Summer 2021   

                                               Producers  \
0                                          Bandai Visual   
1                                 Sunrise, Bandai Visual   
2                                   Victor Entertainment   
3      TV Tokyo, Bandai Visual, Dentsu, Victor Entert...   
4                                       TV Tokyo, Dentsu   
...                                                  ...   
17557                                            Unknown   
17558                                           Kadokawa   
17559                                            Unknown   
17560                                           Kadokawa   
17561                                  Bandai Namco Arts   

                                  Licensors         Studios        Source  \
0          Funimation, Bandai Entertainment         Sunrise  

In [8]:
# Function to classify animes according to their score
animes_df['Appreciation'] = animes_df['Score'].apply(fc.find_rating)

In [9]:
animes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17562 entries, 0 to 17561
Data columns (total 36 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   MAL_ID         17562 non-null  int64  
 1   Name           17562 non-null  object 
 2   Score          12421 non-null  float64
 3   Genres         17562 non-null  object 
 4   English name   17562 non-null  object 
 5   Japanese name  17562 non-null  object 
 6   Type           17562 non-null  object 
 7   Episodes       17562 non-null  int64  
 8   Aired          17562 non-null  object 
 9   Premiered      17562 non-null  object 
 10  Producers      17562 non-null  object 
 11  Licensors      17562 non-null  object 
 12  Studios        17562 non-null  object 
 13  Source         17562 non-null  object 
 14  Duration       17562 non-null  object 
 15  Rating         17562 non-null  object 
 16  Ranked         17562 non-null  object 
 17  Popularity     17562 non-null  int64  
 18  Member

In [10]:
# Taking a look at the best scores
best_animes = animes_df.sort_values(by='Score', ascending=False)
best_animes.head()

MAL_ID                                  Name  Score  \
3971     5114      Fullmetal Alchemist: Brotherhood   9.19   
15926   40028  Shingeki no Kyojin: The Final Season   9.17   
5683     9253                           Steins;Gate   9.11   
14963   38524    Shingeki no Kyojin Season 3 Part 2   9.10   
9913    28977                              Gintama°   9.10   

                                                  Genres  \
3971   Action, Military, Adventure, Comedy, Drama, Ma...   
15926  Action, Military, Mystery, Super Power, Drama,...   
5683                                    Thriller, Sci-Fi   
14963  Action, Drama, Fantasy, Military, Mystery, Sho...   
9913   Action, Comedy, Historical, Parody, Samurai, S...   

                          English name               Japanese name Type  \
3971   Fullmetal Alchemist:Brotherhood  鋼の錬金術師 FULLMETAL ALCHEMIST   TV   
15926     Attack on Titan Final Season      進撃の巨人 The Final Season   TV   
5683                       Steins;Gate                 STEINS;GATE   TV   
14963  Attack on Titan Season 3 Part 2        進撃の巨人 Season3 Part.2   TV   
9913                  Gintama Season 4                         銀魂°   TV   

       Episodes                        Aired    Premiered  \
3971         64   Apr 5, 2009 to Jul 4, 2010  Spring 2009   
15926        16             Dec 7, 2020 to ?  Winter 2021   
5683         24  Apr 6, 2011 to Sep 14, 2011  Spring 2011   
14963        10  Apr 29, 2019 to Jul 1, 2019  Spring 2019   
9913         51  Apr 8, 2015 to Mar 30, 2016  Spring 2015   

                                               Producers  \
3971   Aniplex, Square Enix, Mainichi Broadcasting Sy...   
15926  Production I.G, Dentsu, Mainichi Broadcasting ...   
5683   Frontier Works, Media Factory, Movic, AT-X, Ka...   
14963  Production I.G, Dentsu, Mainichi Broadcasting ...   
9913                           TV Tokyo, Aniplex, Dentsu   

                            Licensors                Studios        Source  \
3971   Funimation, Aniplex of America                  Bones         Manga   
15926                      Funimation                  MAPPA         Manga   
5683                       Funimation              White Fox  Visual novel   
14963                      Funimation             Wit Studio         Manga   
9913          Funimation, Crunchyroll  Bandai Namco Pictures         Manga   

              Duration                          Rating Ranked  Popularity  \
3971   24 min. per ep.  R - 17+ (violence & profanity)    1.0           3   
15926  23 min. per ep.  R - 17+ (violence & profanity)    2.0         119   
5683   24 min. per ep.       PG-13 - Teens 13 or older    3.0           9   
14963  23 min. per ep.  R - 17+ (violence & profanity)    4.0          63   
9913   24 min. per ep.       PG-13 - Teens 13 or older    5.0         329   

       Members  Favorites  Watching  Completed  On-Hold  Dropped  \
3971   2248456     183914    171871    1644938    75728    32456   
15926   733260      44862    566239        553     6658     2090   
5683   1771162     148452    114349    1134756    60926    35600   
14963  1073626      40985     54607     906824     5533     3099   
9913    404121      11868     48011     167130    16612    11472   

       Plan to Watch  Score-10   Score-9   Score-8  Score-7  Score-6  Score-5  \
3971          323463    714811  401507.0  199160.0  70045.0  20210.0   9308.0   
15926         157720    173154   63756.0   26016.0   8793.0   2674.0   1336.0   
5683          425531    468504  275960.0  140914.0  57740.0  21375.0  11126.0   
14963         103563    327290  239451.0  110481.0  33662.0   8365.0   2974.0   
9913          160896     78918   37902.0   21360.0  10215.0   3898.0   2311.0   

      Score-4 Score-3 Score-2  Score-1 Appreciation  
3971   3222.0  1536.0  2162.0    16806    Excellent  
15926   588.0   382.0   514.0    11061    Excellent  
5683   5061.0  2292.0  1678.0     5255    Excellent  
14963  1108.0   550.0   385.0     4169    Excellen

In [36]:
# Using the get_genres functions to split the Genres column and get unique_genres.
df_copy = pc.split_genres(animes_df)
unique_genres = pc.get_genres(animes_df)


In [12]:
order = ['Excellent', 'Good', 'Bad', 'Terrible']
fig1 = px.histogram(df_copy.explode('Genres'), x='Appreciation', color='Genres', category_orders={'Genres': unique_genres}, title='Appreciation by Genre', 
                    color_discrete_map={'Action': 'DeepSkyBlue', 'Adventure': 'DodgerBlue', 'Cars': 'RoyalBlue', 'Comedy': 'Blue', 
                                        'Dementia': 'DarkBlue', 'Demons': 'MidnightBlue', 'Drama': 'SlateBlue', 'Ecchi': 'Cyan', 
                                       'Fantasy': 'Turquoise', 'Game': 'DarkCyan', 'Harem': 'MediumAquamarine', 'Hentai': 'MediumSpringGreen', 
                                       'Historical': 'SpringGreen', 'Horror': 'LimeGreen', 'Josei': 'Green', 'Kids': 'DarkGreen', 
                                       'Magic': 'YellowGreen', 'Martial Arts': 'GreenYellow', 'Mecha': 'Goldenrod', 'Military': 'DarkGoldenrod', 
                                       'Music': 'SandyBrown', 'Mystery': 'Peru', 'Parody': 'Chocolate', 'Police': 'LightPink', 
                                       'Psychological': 'LightCoral', 'Romance': 'Crimson', 'Samurai': 'DarkRed', 'School': 'Red', 
                                       'Sci-Fi': 'Tomato', 'Seinen': 'OrangeRed', 'Shoujo': 'DarkOrange', 'Shoujo Ai': 'Orange', 
                                       'Shounen': 'Gold', 'Shounen Ai': 'Yellow', 'Slice of Life': 'Khaki', 'Space': 'LightGoldenrodYellow', 
                                       'Sports': 'PaleGoldenrod', 'Super Power': 'LightCyan', 'Supernatural': 'PowderBlue',
                                        'Thriller': 'PaleTurquoise', 'Unknown': 'MintCream', 'Vampire': 'Seashell', 'Yaoi': 'OldLace', 
                                       'Yuri': 'Beige'})
fig1.update_xaxes(categoryorder='array', categoryarray=order)

fig1.show()

We can see that there are very few excellent animes. How about taking a look at them?

In [13]:
# Defining the columns
ex_cols = ['Name', 'Score', 'Genres', 'Type', 'Episodes', 'Appreciation']
# Extracting only the excellent animes
excellents = animes_df[animes_df['Appreciation'] == 'Excellent'][ex_cols]
# Sorting based on score
excellents = excellents.sort_values(by='Score', ascending=False)
excellents

Name  Score  \
3971       Fullmetal Alchemist: Brotherhood   9.19   
15926  Shingeki no Kyojin: The Final Season   9.17   
5683                            Steins;Gate   9.11   
6474                 Hunter x Hunter (2011)   9.10   
9913                               Gintama°   9.10   
14963    Shingeki no Kyojin Season 3 Part 2   9.10   
6006                               Gintama'   9.08   
741                    Ginga Eiyuu Densetsu   9.07   
7261                    Gintama': Enchousen   9.04   
9886                         Koe no Katachi   9.00   
12898            3-gatsu no Lion 2nd Season   9.00   

                                                  Genres   Type  Episodes  \
3971   Action, Military, Adventure, Comedy, Drama, Ma...     TV        64   
15926  Action, Military, Mystery, Super Power, Drama,...     TV        16   
5683                                    Thriller, Sci-Fi     TV        24   
6474    Action, Adventure, Fantasy, Shounen, Super Power     TV       148   
9913   Action, Comedy, Historical, Parody, Samurai, S...     TV        51   
14963  Action, Drama, Fantasy, Military, Mystery, Sho...     TV        10   
6006   Action, Sci-Fi, Comedy, Historical, Parody, Sa...     TV        51   
741                       Military, Sci-Fi, Space, Drama    OVA       110   
7261   Action, Comedy, Historical, Parody, Samurai, S...     TV        13   
9886                              Drama, School, Shounen  Movie         1   
12898                 Drama, Game, Seinen, Slice of Life     TV        22   

      Appreciation  
3971     Excellent  
15926    Excellent  
5683     Excellent  
6474     Excellent  
9913     Excellent  
14963    Excellent  
6006     Excellent  
741      Excellent  
7261     Excellent  
9886     Excellent  
12898    Excellent

These are all the animes with a rating higher than 9. There are really few of them, I think they all must be amazing.

### Most completed animes

In [14]:
comp_cols = ['Name', 'Score', 'Genres', 'Type', 'Episodes', 'Completed', 'Appreciation']
most_completed = animes_df.sort_values(by='Completed', ascending=False)[comp_cols]
most_completed = most_completed.head(10)
most_completed

Name  Score  \
7449                 Shingeki no Kyojin   8.48   
1393                         Death Note   8.63   
6614                   Sword Art Online   7.25   
10451                     One Punch Man   8.57   
11185             Boku no Hero Academia   8.11   
3971   Fullmetal Alchemist: Brotherhood   9.19   
8646                        Tokyo Ghoul   7.81   
10                               Naruto   7.91   
11308                    Kimi no Na wa.   8.96   
8148                    No Game No Life   8.20   

                                                  Genres   Type  Episodes  \
7449   Action, Military, Mystery, Super Power, Drama,...     TV        25   
1393   Mystery, Police, Psychological, Supernatural, ...     TV        37   
6614           Action, Game, Adventure, Romance, Fantasy     TV        25   
10451  Action, Sci-Fi, Comedy, Parody, Super Power, S...     TV        12   
11185       Action, Comedy, School, Shounen, Super Power     TV        13   
3971   Action, Military, Adventure, Comedy, Drama, Ma...     TV        64   
8646   Action, Mystery, Horror, Psychological, Supern...     TV        12   
10     Action, Adventure, Comedy, Super Power, Martia...     TV       220   
11308               Romance, Supernatural, School, Drama  Movie         1   
8148   Game, Adventure, Comedy, Supernatural, Ecchi, ...     TV        12   

       Completed Appreciation  
7449     2182587         Good  
1393     2146116         Good  
6614     1907261         Good  
10451    1841220         Good  
11185    1655900         Good  
3971     1644938    Excellent  
8646     1594880         Good  
10       1462223         Good  
11308    1462143         Good  
8148     1426896         Good

These are the 10 most watched animes. A lot of people saw Shingeki no Kyojin huh. I bet they are all worth watching.

### Most dropped animes

In [15]:
drop_cols = ['Name', 'Score', 'Genres', 'Type', 'Episodes', 'Dropped', 'Appreciation']
most_dropped = animes_df.sort_values(by='Dropped', ascending=False)[drop_cols]
most_dropped = most_dropped.head(10)
most_dropped

Name  Score  \
245                             Bleach   7.80   
4707                        Fairy Tail   7.68   
11                           One Piece   8.52   
1574                Naruto: Shippuuden   8.16   
12492  Boruto: Naruto Next Generations   5.81   
10                              Naruto   7.91   
6614                  Sword Art Online   7.25   
12493                     Black Clover   7.38   
1393                        Death Note   8.63   
7939                      Kill la Kill   8.11   

                                                  Genres Type  Episodes  \
245    Action, Adventure, Comedy, Super Power, Supern...   TV       366   
4707   Action, Adventure, Comedy, Magic, Fantasy, Sho...   TV       175   
11     Action, Adventure, Comedy, Super Power, Drama,...   TV         0   
1574   Action, Adventure, Comedy, Super Power, Martia...   TV       500   
12492  Action, Adventure, Super Power, Martial Arts, ...   TV         0   
10     Action, Adventure, Comedy, Super Power, Martia...   TV       220   
6614           Action, Game, Adventure, Romance, Fantasy   TV        25   
12493            Action, Comedy, Magic, Fantasy, Shounen   TV       170   
1393   Mystery, Police, Psychological, Supernatural, ...   TV        37   
7939          Action, Comedy, Super Power, Ecchi, School   TV        24   

       Dropped Appreciation  
245     174710         Good  
4707    148408         Good  
11      136245         Good  
1574    124253         Good  
12492   113677          Bad  
10       99806         Good  
6614     90661         Good  
12493    89594         Good  
1393     80834         Good  
7939     67845         Good

These are the 10 most dropped animes, there are a lot of anime with good appreciation here. I wonder why did so many people drop it.

### Most favorite animes

In [16]:
fav_cols = ['Name', 'Score', 'Genres', 'Type', 'Episodes', 'Favorites', 'Appreciation']
most_favorite = animes_df.sort_values(by='Favorites', ascending=False)[fav_cols]
most_favorite = most_favorite.head(10)
most_favorite

Name  Score  \
3971   Fullmetal Alchemist: Brotherhood   9.19   
5683                        Steins;Gate   9.11   
6474             Hunter x Hunter (2011)   9.10   
1393                         Death Note   8.63   
7449                 Shingeki no Kyojin   8.48   
11                            One Piece   8.52   
1431    Code Geass: Hangyaku no Lelouch   8.72   
1574                 Naruto: Shippuuden   8.16   
20              Neon Genesis Evangelion   8.32   
11308                    Kimi no Na wa.   8.96   

                                                  Genres   Type  Episodes  \
3971   Action, Military, Adventure, Comedy, Drama, Ma...     TV        64   
5683                                    Thriller, Sci-Fi     TV        24   
6474    Action, Adventure, Fantasy, Shounen, Super Power     TV       148   
1393   Mystery, Police, Psychological, Supernatural, ...     TV        37   
7449   Action, Military, Mystery, Super Power, Drama,...     TV        25   
11     Action, Adventure, Comedy, Super Power, Drama,...     TV         0   
1431   Action, Military, Sci-Fi, Super Power, Drama, ...     TV        25   
1574   Action, Adventure, Comedy, Super Power, Martia...     TV       500   
20     Action, Sci-Fi, Dementia, Psychological, Drama...     TV        26   
11308               Romance, Supernatural, School, Drama  Movie         1   

       Favorites Appreciation  
3971      183914    Excellent  
5683      148452    Excellent  
6474      147274    Excellent  
1393      145201         Good  
7449      129844         Good  
11        126645         Good  
1431       90487         Good  
1574       84651         Good  
20         71308         Good  
11308      71054         Good

Would these be the 10 animes most loved by fans? It's worth giving such beloved animes a chance.

### Animes with the most 10 ratings

In [17]:
s10_cols = ['Name', 'Score', 'Genres', 'Type', 'Episodes', 'Score-10', 'Appreciation']
most_score_10 = animes_df.sort_values(by='Score-10', ascending=False)[s10_cols]
most_score_10 = most_score_10.head(10)
most_score_10

Name  Score  \
3971     Fullmetal Alchemist: Brotherhood   9.19   
1393                           Death Note   8.63   
11308                      Kimi no Na wa.   8.96   
7449                   Shingeki no Kyojin   8.48   
5683                          Steins;Gate   9.11   
6474               Hunter x Hunter (2011)   9.10   
9886                       Koe no Katachi   9.00   
10451                       One Punch Man   8.57   
2656   Code Geass: Hangyaku no Lelouch R2   8.91   
14963  Shingeki no Kyojin Season 3 Part 2   9.10   

                                                  Genres   Type  Episodes  \
3971   Action, Military, Adventure, Comedy, Drama, Ma...     TV        64   
1393   Mystery, Police, Psychological, Supernatural, ...     TV        37   
11308               Romance, Supernatural, School, Drama  Movie         1   
7449   Action, Military, Mystery, Super Power, Drama,...     TV        25   
5683                                    Thriller, Sci-Fi     TV        24   
6474    Action, Adventure, Fantasy, Shounen, Super Power     TV       148   
9886                              Drama, School, Shounen  Movie         1   
10451  Action, Sci-Fi, Comedy, Parody, Super Power, S...     TV        12   
2656   Action, Military, Sci-Fi, Super Power, Drama, ...     TV        25   
14963  Action, Drama, Fantasy, Military, Mystery, Sho...     TV        10   

       Score-10 Appreciation  
3971     714811    Excellent  
1393     557406         Good  
11308    516874         Good  
7449     470882         Good  
5683     468504    Excellent  
6474     454000    Excellent  
9886     393684    Excellent  
10451    360187         Good  
2656     358705         Good  
14963    327290    Excellent

These are the animes with the most votes out of 10. I bet these are animes that had a huge impact on all these people.

### Animes with the most 1 ratings

In [18]:
s1_cols = ['Name', 'Score', 'Genres', 'Type', 'Episodes', 'Score-1', 'Appreciation']
most_score_1 = animes_df.sort_values(by='Score-1', ascending=False)[s1_cols]
most_score_1 = most_score_1.head(10)
most_score_1

Name  Score  \
1490                       Boku no Pico   4.37   
2271                        School Days   5.68   
8058                               Pupa   3.41   
6614                   Sword Art Online   7.25   
387                     Hametsu no Mars   2.23   
3971   Fullmetal Alchemist: Brotherhood   9.19   
8551                Sword Art Online II   6.79   
15172                            Ex-Arm   2.18   
12492   Boruto: Naruto Next Generations   5.81   
2934       Tenkuu Danzai Skelter+Heaven   1.85   

                                                  Genres Type  Episodes  \
1490                                        Hentai, Yaoi  OVA         1   
2271                       Harem, Drama, Romance, School   TV        12   
8058                      Fantasy, Horror, Psychological   TV        12   
6614           Action, Game, Adventure, Romance, Fantasy   TV        25   
387                                       Sci-Fi, Horror  OVA         1   
3971   Action, Military, Adventure, Comedy, Drama, Ma...   TV        64   
8551           Action, Game, Adventure, Romance, Fantasy   TV        24   
15172                      Action, Sci-Fi, Ecchi, Seinen   TV        12   
12492  Action, Adventure, Super Power, Martial Arts, ...   TV         0   
2934                                       Sci-Fi, Mecha  OVA         1   

       Score-1 Appreciation  
1490     33750          Bad  
2271     25046          Bad  
8058     23622          Bad  
6614     23472         Good  
387      22048     Terrible  
3971     16806    Excellent  
8551     16529         Good  
15172    15346     Terrible  
12492    13039          Bad  
2934     11816     Terrible

These are the animes with the most votes out of 1. I bet these are animes that had a huge impact on all these people, but... in the bad way.

### Some graphs with interesting data

In [19]:
fig2 = px.histogram(df_copy.explode('Genres'), x='Score', color='Genres', category_orders={'Genres': unique_genres}, title='Distribution of Anime Score', 
                   color_discrete_map={'Action': 'DeepSkyBlue', 'Adventure': 'DodgerBlue', 'Cars': 'RoyalBlue', 'Comedy': 'Blue', 
                                        'Dementia': 'DarkBlue', 'Demons': 'MidnightBlue', 'Drama': 'SlateBlue', 'Ecchi': 'Cyan', 
                                       'Fantasy': 'Turquoise', 'Game': 'DarkCyan', 'Harem': 'MediumAquamarine', 'Hentai': 'MediumSpringGreen', 
                                       'Historical': 'SpringGreen', 'Horror': 'LimeGreen', 'Josei': 'Green', 'Kids': 'DarkGreen', 
                                       'Magic': 'YellowGreen', 'Martial Arts': 'GreenYellow', 'Mecha': 'Goldenrod', 'Military': 'DarkGoldenrod', 
                                       'Music': 'SandyBrown', 'Mystery': 'Peru', 'Parody': 'Chocolate', 'Police': 'LightPink', 
                                       'Psychological': 'LightCoral', 'Romance': 'Crimson', 'Samurai': 'DarkRed', 'School': 'Red', 
                                       'Sci-Fi': 'Tomato', 'Seinen': 'OrangeRed', 'Shoujo': 'DarkOrange', 'Shoujo Ai': 'Orange', 
                                       'Shounen': 'Gold', 'Shounen Ai': 'Yellow', 'Slice of Life': 'Khaki', 'Space': 'LightGoldenrodYellow', 
                                       'Sports': 'PaleGoldenrod', 'Super Power': 'LightCyan', 'Supernatural': 'PowderBlue',
                                        'Thriller': 'PaleTurquoise', 'Unknown': 'MintCream', 'Vampire': 'Seashell', 'Yaoi': 'OldLace', 
                                       'Yuri': 'Beige'})
fig2.show()

In [38]:
# Using the explode_genres function to split and explode the Genres column.
df_copy = pc.explode_genres(animes_df)
# Anime count by genre
genre_counts = df_copy['Genres'].value_counts()

In [29]:
fig3 = px.bar(x=genre_counts.index, y=genre_counts.values, labels={'x': 'Genre', 'y': 'Count'}, 
             title='Number of Animes by Genre', color_discrete_sequence=['#386df9'])

fig3.show()

In [39]:
# Using the explode_genres function to split and explode the Genres column.
df_copy = pc.explode_genres(animes_df)
# Grouping by genres and adding completed animes.
completed_genres = df_copy.groupby('Genres')['Completed'].sum().reset_index()
completed_genres = completed_genres.sort_values(by='Completed', ascending=False)

In [40]:
fig4 = px.bar(completed_genres, x='Genres', y='Completed', title='Most Watched Genres', 
              color_discrete_sequence=['#386df9'])

fig4.show()

In [32]:
# Calculating the average score for each gender
genre_mean_score = df_copy.groupby('Genres')['Score'].mean().reset_index()

In [33]:
fig5 = px.bar(genre_mean_score, x='Genres', y='Score', title='Average Score by Genre', 
              color_discrete_sequence=['#386df9'] * len(genre_mean_score))

fig5.show()

In [25]:
fig6 = px.scatter(animes_df, x='Score', y='Favorites', title='Relation Between Score and Favorites', 
                  color_discrete_sequence=['#386df9'])

fig6.show()

In [26]:
# Counting the number of each type of anime
type_counts = animes_df['Type'].value_counts().reset_index()
# Renaming the columns
type_counts.columns = ['Type', 'Count']

In [27]:
fig7 = px.bar(type_counts, x='Type', y='Count', title='Distribution of Animes by Type', 
             color_discrete_sequence=['#386df9'])

fig7.show()

### Conclusions

After separating the animes by scores, we see that the majority of animes are considered good (Score greater than 6), but only 11 animes are considered excellent (Score greater than 9). Of these 11 excellent animes, 7 of them are in the action genre, 9 of them are in the TV genre, and 8 of them have 51 or fewer episodes.

In the 10 most watched animes we see that only one of them has a score higher than 9, but the minimum score among them is 7.25, we see the action genre present in 7 animes, and we also see 8 animes with 37 or fewer episodes. We can already see great relevance in the action genre and in animes with few episodes.

In the most dropped animes we see that 9 of the 10 have a score greater than 7, and that 7 of these animes have many episodes (between 170 and 500), 2 of them are still being broadcast, we can conclude that animes the public has a greater tendency to drop animes with many episodes.

In the 10 animes with the most favorites, we maintain the trend with 7 animes having the action genre and having few episodes, however we found some exceptions, such as an anime with a score of 9.11 having a thriller and scifi genre, an anime with 500 episodes, and also an anime still broadcasting.

In animes with more 10-Score we maintain the trend of the action genre and few episodes, but with more exceptions in the genre, such as mystery, romance, drama, school, and thriller.

When analyzing the distribution of scores, we noticed that most scores are between 6 and 7.5.

Analyzing the most watched genres, we noticed a great advantage in action and comedy, which are also the genres that appear the most. We noticed some other popular genres as well, such as Shonen, Drama, Romance and Fantasy.

However, when analyzing the average score for each genre, we noticed all genres with a very similar average ranging between 6 and 7.

### Final Conclusions

After all the analysis, we can conclude that the best safe choices to launch or announce a new anime are animes containing the Action or Comedy genre, which may contain subgenres such as Shonen or Drama, and have few episodes so that more people can watch. If you are working with films, Romance and Drama films are best rated. Some more distinct genres can also be successful, such as Thriller, Scifi and Mystery, as well as long animes can also be exceptions, being in the Shonen genre.